### PR TITLE
Add sync events created after an update to a job run

### PR DESCRIPTION
**Pull request by j16r**
Add a callback system to record events, then a simple stats pusher which feeds them to the WebSocketClient.

Renamed StatsPusher to WebSocketClient.

### CODE REVIEW

File: store/models/events.go [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package models
+
+import "time"
+
+// SyncEvent represents an event sourcing style event, which is used to sync
+// data upstream with another service
+type SyncEvent struct {
+	ID        int       `json:"id" gorm:"primary_key"`
```
Comment (dimroc) [inline store/models/events.go]: optional: `gorm.Model` makes this as a `uint`, we could follow their lead.

File: integration/features_test.go [hunk 2]
```diff
@@ -616,3 +616,34 @@ func TestIntegration_CreateServiceAgreement(t *testing.T) {
 	eth.EventuallyAllCalled(t)
 
 }
+
+func TestIntegration_SyncJobRuns(t *testing.T) {
+	wsserver, wsserverCleanup := cltest.NewEventWebSocketServer(t)
+	defer wsserverCleanup()
+
+	t.Parallel()
```
Comment (dimroc) [inline integration/features_test.go]: Should this be at the top of the test?

File: store/orm/orm.go [hunk 3]
```diff
@@ -66,10 +67,37 @@ func NewORM(uri string, timeout time.Duration) (*ORM, error) {
...
+
+	if scope.TableName() == "job_runs" {
+		bodyBytes, err := json.Marshal(scope.Value)
+		if err != nil {
+			scope.Err(err)
+			return
+		}
+
+		event := models.SyncEvent{
+			Body: string(bodyBytes),
```
Comment (dimroc) [inline store/orm/orm.go]: Am I correct in assuming this wouldn't include updates to task runs or any details on associations? Seems fine as a first step, but we would need to expand this eventually. You wouldn't happen to have an example of the body string that's saved would you?

File: store/orm/orm.go [hunk 4]
```diff
@@ -66,10 +67,37 @@ func NewORM(uri string, timeout time.Duration) (*ORM, error) {
...
-	return &ORM{
+
+	orm := &ORM{
 		DB:              db,
 		lockingStrategy: lockingStrategy,
-	}, nil
+	}
+
+	if linkstatsURL != nil {
+		db.Callback().Create().After("gorm:update").Register("sync:run_after_create", createSyncEvents)
```
Comment (dimroc) [inline store/orm/orm.go]: Personally, I would prefer to keep the ORM focused on the single responsibility of wrapping the db persistence layer. Could we pull this up just one layer to the `store` and have a separate type (`StatsPusher` or a new `SyncEventManager`?) set up the callback `createSyncEvents` via the public `DB`? This way, if the logic in `createSyncEvents` grows, we don't intrude on `orm.go`.

That being said, perhaps you see sync events as part of ORM.

File: store/stats_pusher.go [hunk 5]
```diff
@@ -2,229 +2,84 @@ package store
...
+		eq.cancel()
 	}
+	return eq.WSClient.Close()
 }
 
-var (
-	newline = []byte{'\n'}
-)
+func (eq *StatsPusher) pollEvents(parentCtx context.Context) {
+	pollTicker := time.NewTicker(eq.Period)
```
Comment (dimroc) [inline store/stats_pusher.go]: Not for this PR as discussed but I'm excited to see if we can elegantly get off polling. There are implications with integration testing if we have to wait one minute for a sync event.

File: store/orm/orm.go (hunk 3 above)
Comment (j16r) [reply store/orm/orm.go]: `{"id":"d19e1df47ecb40fa85e7f29b4c25cd6e","jobId":"b7dbc97018ce4652b79f3f17e20fce00","result":{"jobRunId":"","taskRunId":"","data":{},"status":"","error":null},"status":"in_progress","taskRuns":[{"id":"c4067457f59b4f3d88dda39c80178995","result":{"jobRunId":"d19e1df47ecb40fa85e7f29b4c25cd6e","taskRunId":"c4067457f59b4f3d88dda39c80178995","data":{},"status":"","error":null},"status":"","task":{"ID":1,"CreatedAt":"2019-03-25T12:33:34.955039-07:00","UpdatedAt":"2019-03-25T12:33:34.955039-07:00","DeletedAt":null,"type":"noop","confirmations":0,"params":{}},"minimumConfirmations":0}],"createdAt":"2019-03-25T12:33:34.956255-07:00","completedAt":null,"updatedAt":"2019-03-25T12:33:34.956255-07:00","initiator":{"id":1,"jobSpecId":"b7dbc97018ce4652b79f3f17e20fce00","type":"runat","CreatedAt":"2019-03-25T12:33:34.954949-07:00","params":{"time":"2018-01-08T18:12:01.103Z","ran":true,"address":"0x0000000000000000000000000000000000000000"}},"creationHeight":null,"observedHeight":null,"overrides":{"jobRunId":"","taskRunId":"","data":{},"status":"","error":null}}`

File: store/orm/orm.go (hunk 4 above)
Comment (j16r) [reply store/orm/orm.go]: I've put the callback into another file, to separate it.

File: store/stats_pusher.go [hunk 6]
```diff
@@ -2,229 +2,84 @@ package store
...
-		return fmt.Errorf("websocketStatsPusher#connect: %v", err)
-	}
-
-	w.conn = conn
-	return nil
-}
+				eq.WSClient.Send([]byte(event.Body))
 
-var expectedCloseMessages = []int{websocket.CloseGoingAway, websocket.CloseAbnormalClosure, websocket.CloseNormalClosure}
+				// TODO: This is fire and forget, we may want to get confirmation
```
Comment (se3000) [inline store/stats_pusher.go]: https://www.pivotaltracker.com/story/show/164879382

File: store/orm/orm.go (hunk 3 above)
Comment (dimroc) [reply store/orm/orm.go]: Has `taskRuns :[...]` 👍 

Will probably need to scrub that of private info in a future story.

File: store/orm/orm.go (hunk 4 above)
Comment (dimroc) [reply store/orm/orm.go]: Hey there, don't fully understand where the complexity comes in. The `scope.DB` belongs to `gorm` which can easily get passed in the `store` package or any other package without a circular dependency. Couldn't you just put this (and `createSyncEvents`) in the `StatsPusher (store/stats_pusher.go)` constructor with little hassle?

```golang
// NewStatsPusher returns a new event queuer
func NewStatsPusher(orm *orm.ORM, url *url.URL) *StatsPusher {
	var wsClient WebSocketClient
	wsClient = noopWebSocketClient{}
	if url != nil {
		wsClient = NewWebSocketClient(url)
                orm.DB.Callback().Create().After("gorm:update").Register("sync:run_after_create", createSyncEvents)
	}
```

Perhaps it's a reluctance to place `gorm` outside of `orm`. Will sync with you offline.

**Pull request merged**